File: backend-nodejs-backup/src/controllers/message.controller.js
import cloudinary from "../lib/cloudinary.js";
import { getReceiverSocketId, io } from "../lib/socket.js";
import Message from "../models/Message.js";
import User from "../models/User.js";
import axios from "axios";

export const getAllContacts = async (req, res) => {
  try {
    const loggedInUserId = req.user._id;
    const filteredUsers = await User.find({ _id: { $ne: loggedInUserId } }).select("-password");

    res.status(200).json(filteredUsers);
  } catch (error) {
    console.log("Error in getAllContacts:", error);
    res.status(500).json({ message: "Server error" });
  }
};

export const getMessagesByUserId = async (req, res) => {
  try {
    const myId = req.user._id;
    const { id: userToChatId } = req.params;

    const messages = await Message.find({
      $or: [
        { senderId: myId, receiverId: userToChatId },
        { senderId: userToChatId, receiverId: myId },
      ],
    });

    const decryptedMessages = await Promise.all(
      messages.map(async (msg) => {
        const msgObj = msg.toObject();
        if (msgObj.isEncrypted) {
          try {
            const decryptRes = await axios.post(`${process.env.CRYPTO_SERVICE_URL}/decrypt`, {
              from: msgObj.senderId.toString(),
              to: msgObj.receiverId.toString(),
              ciphertext: msgObj.ciphertext,
              messageType: msgObj.messageType,
              sessionId: msgObj.sessionId,
            });
            msgObj.text = decryptRes.data.plaintext;
          } catch (error) {
            console.error("Decryption error:", error.message);
            msgObj.text = "[Cannot decrypt]";
          }
        }
        return msgObj;
      })
    );

    res.status(200).json(decryptedMessages);
  } catch (error) {
    console.log("Error in getMessages controller: ", error.message);
    res.status(500).json({ error: "Internal server error" });
  }
};

export const sendMessage = async (req, res) => {
  try {
    const { text, image } = req.body;
    const { id: receiverId } = req.params;
    const senderId = req.user._id;

    if (!text && !image) {
      return res.status(400).json({ message: "Text or image is required." });
    }
    if (senderId.equals(receiverId)) {
      return res.status(400).json({ message: "Cannot send messages to yourself." });
    }
    const receiverExists = await User.exists({ _id: receiverId });
    if (!receiverExists) {
      return res.status(404).json({ message: "Receiver not found." });
    }

    let imageUrl;
    if (image) {
      // upload base64 image to cloudinary
      const uploadResponse = await cloudinary.uploader.upload(image);
      imageUrl = uploadResponse.secure_url;
    }

    let isEncrypted = false;
    let cryptoData = {};

    if (text) {
      let recipientBundle;
      try {
        const bundleRes = await axios.get(`${process.env.CRYPTO_SERVICE_URL}/bundle/${receiverId}`);
        recipientBundle = bundleRes.data;
      } catch (error) {
        await axios.post(`${process.env.CRYPTO_SERVICE_URL}/generate-keys`, { userId: receiverId });
        const bundleRes = await axios.get(`${process.env.CRYPTO_SERVICE_URL}/bundle/${receiverId}`);
        recipientBundle = bundleRes.data;
      }

      try {
        await axios.get(`${process.env.CRYPTO_SERVICE_URL}/bundle/${senderId}`);
      } catch (error) {
        await axios.post(`${process.env.CRYPTO_SERVICE_URL}/generate-keys`, { userId: senderId });
      }

      const encryptRes = await axios.post(`${process.env.CRYPTO_SERVICE_URL}/encrypt`, {
        from: senderId.toString(),
        to: receiverId.toString(),
        plaintext: text,
        recipientBundle,
      });

      cryptoData = {
        ciphertext: encryptRes.data.ciphertext,
        messageType: encryptRes.data.messageType,
        sessionId: encryptRes.data.sessionId,
      };
      isEncrypted = true;
    }

    const newMessageData = {
      senderId,
      receiverId,
      image: imageUrl,
    };

    if (isEncrypted) {
      newMessageData.isEncrypted = true;
      newMessageData.ciphertext = cryptoData.ciphertext;
      newMessageData.messageType = cryptoData.messageType;
      newMessageData.sessionId = cryptoData.sessionId;
    } else {
      newMessageData.text = text;
    }

    const newMessage = new Message(newMessageData);
    await newMessage.save();

    const messageForSocket = newMessage.toObject();
    if (isEncrypted) {
      messageForSocket.text = text;
      delete messageForSocket.ciphertext;
      delete messageForSocket.messageType;
      delete messageForSocket.sessionId;
    }

    const receiverSocketId = getReceiverSocketId(receiverId);
    if (receiverSocketId) {
      io.to(receiverSocketId).emit("newMessage", messageForSocket);
    }

    res.status(201).json(messageForSocket);
  } catch (error) {
    console.log("Error in sendMessage controller: ", error.message);
    res.status(500).json({ error: "Internal server error" });
  }
};

export const getChatPartners = async (req, res) => {
  try {
    const loggedInUserId = req.user._id;

    // find all the messages where the logged-in user is either sender or receiver
    const messages = await Message.find({
      $or: [{ senderId: loggedInUserId }, { receiverId: loggedInUserId }],
    });

    const chatPartnerIds = [
      ...new Set(
        messages.map((msg) =>
          msg.senderId.toString() === loggedInUserId.toString()
            ? msg.receiverId.toString()
            : msg.senderId.toString()
        )
      ),
    ];

    const chatPartners = await User.find({ _id: { $in: chatPartnerIds } }).select("-password");

    res.status(200).json(chatPartners);
  } catch (error) {
    console.error("Error in getChatPartners: ", error.message);
    res.status(500).json({ error: "Internal server error" });
  }
};

File: backend/src/controllers/message_controller.py
from bson import ObjectId
from datetime import datetime
from src.lib.db import get_db
from src.lib.cloudinary import upload_image
from src.lib.socket import get_receiver_socket_id, emit_new_message, emit_online_users

async def get_all_contacts(user_id: str):
    try:
        db = get_db()
        if not ObjectId.is_valid(user_id):
            return ({'error': 'Invalid user ID'}, 400)
        users = await db['users'].find({'_id': {'$ne': ObjectId(user_id)}}, {'password': 0}).to_list(None)
        for user in users:
            user['_id'] = str(user['_id'])
        return (users, 200)
    except Exception as e:
        print(f'Error in get_all_contacts: {e}')
        return ({'error': 'Server error'}, 500)

async def get_messages_by_user_id(my_id: str, user_to_chat_id: str):
    try:
        db = get_db()
        if not ObjectId.is_valid(my_id) or not ObjectId.is_valid(user_to_chat_id):
            return ({'error': 'Invalid user ID'}, 400)
        my_id_obj = ObjectId(my_id)
        user_to_chat_id_obj = ObjectId(user_to_chat_id)
        messages = await db['messages'].find({'$or': [{'senderId': my_id_obj, 'receiverId': user_to_chat_id_obj}, {'senderId': user_to_chat_id_obj, 'receiverId': my_id_obj}]}).to_list(None)
        for message in messages:
            message['_id'] = str(message['_id'])
            message['senderId'] = str(message['senderId'])
            message['receiverId'] = str(message['receiverId'])
            if isinstance(message.get('createdAt'), datetime):
                message['createdAt'] = message['createdAt'].isoformat()
        return (messages, 200)
    except Exception as e:
        print(f'Error in get_messages_by_user_id: {e}')
        return ({'error': 'Internal server error'}, 500)

async def send_message(sender_id: str, receiver_id: str, ciphertext: str=None, message_type: int=None, session_id: str=None, image: str=None):
    try:
        print(f'[Message] Relaying message from {sender_id} to {receiver_id}')
        db = get_db()
        if not ciphertext and (not image):
            return ({'error': 'Ciphertext or image is required.'}, 400)
        if not ObjectId.is_valid(sender_id) or not ObjectId.is_valid(receiver_id):
            return ({'error': 'Invalid user ID'}, 400)
        sender_id_obj = ObjectId(sender_id)
        receiver_id_obj = ObjectId(receiver_id)
        if sender_id_obj == receiver_id_obj:
            return ({'error': 'Cannot send messages to yourself.'}, 400)
        receiver_exists = await db['users'].find_one({'_id': receiver_id_obj})
        if not receiver_exists:
            return ({'error': 'Receiver not found.'}, 404)
        image_url = None
        if image:
            try:
                image_url = upload_image(image)
            except Exception as e:
                print(f'Error uploading image: {e}')
                return ({'error': 'Failed to upload image'}, 500)
        now = datetime.now()
        new_message = {'senderId': sender_id_obj, 'receiverId': receiver_id_obj, 'ciphertext': ciphertext, 'messageType': message_type, 'sessionId': session_id, 'image': image_url, 'createdAt': now}
        result = await db['messages'].insert_one(new_message)
        if result.inserted_id:
            response_message = {'_id': str(result.inserted_id), 'senderId': str(sender_id_obj), 'receiverId': str(receiver_id_obj), 'ciphertext': ciphertext, 'messageType': message_type, 'sessionId': session_id, 'image': image_url, 'createdAt': now.isoformat()}
            await emit_new_message(receiver_id, response_message)
            return (response_message, 201)
        else:
            return ({'error': 'Failed to store message'}, 400)
    except Exception as e:
        print(f'Error in send_message controller: {e}')
        return ({'error': 'Internal server error'}, 500)

async def get_chat_partners(user_id: str):
    try:
        db = get_db()
        if not ObjectId.is_valid(user_id):
            return ({'error': 'Invalid user ID'}, 400)
        user_id_obj = ObjectId(user_id)
        messages = await db['messages'].find({'$or': [{'senderId': user_id_obj}, {'receiverId': user_id_obj}]}).sort('createdAt', -1).to_list(None)
        chat_partners_map = {}
        for msg in messages:
            if msg['senderId'] == user_id_obj:
                partner_id = msg['receiverId']
            else:
                partner_id = msg['senderId']
            partner_id_str = str(partner_id)
            if partner_id_str not in chat_partners_map:
                chat_partners_map[partner_id_str] = msg.get('createdAt')
        chat_partner_ids_obj = [ObjectId(partner_id) for partner_id in chat_partners_map.keys()]
        chat_partners = await db['users'].find({'_id': {'$in': chat_partner_ids_obj}}, {'password': 0}).to_list(None)
        result = []
        for partner in chat_partners:
            partner_id_str = str(partner['_id'])
            partner['_id'] = partner_id_str
            partner['lastMessageDate'] = chat_partners_map[partner_id_str]
            result.append(partner)
        result.sort(key=lambda x: x.get('lastMessageDate') or datetime.utcnow(), reverse=True)
        return (result, 200)
    except Exception as e:
        print(f'Error in get_chat_partners: {e}')
        return ({'error': 'Internal server error'}, 500)

async def get_chat_partner_ids(user_id: str):
    try:
        db = get_db()
        if not ObjectId.is_valid(user_id):
            return ({'error': 'Invalid user ID'}, 400)
        user_id_obj = ObjectId(user_id)
        messages = await db['messages'].find({'$or': [{'senderId': user_id_obj}, {'receiverId': user_id_obj}]}).to_list(None)
        chat_partner_ids = set()
        for msg in messages:
            if msg['senderId'] == user_id_obj:
                chat_partner_ids.add(str(msg['receiverId']))
            else:
                chat_partner_ids.add(str(msg['senderId']))
        result = sorted(list(chat_partner_ids))
        return (result, 200)
    except Exception as e:
        print(f'Error get_chat_partner_ids failed: {e}')
        return ({'error': 'Internal server error'}, 500)